include/crypto/event.h

```diff
@@ -0,0 +1,378 @@
+     *
+     * The functionality this implements is open ended, and allows
+     * extraordinary operations such as reordering (rescheduling) events,
+     * affecting other associated queues, as well as modifications or access
+     * of the auxialliary method data, etc etc etc.
+     * Use with care!
+     *
+     * The arguments are of the same style as ioctl.h functions.  The caller
+     * must know exactly what arguments are expected for each |request|.
+     *
+     * Expected to return 1 on success, -1 on failure.
+     */
+    int (*ctrl)(struct ossl_event_queue_st *queue, uint32_t request,
+                ... /* args */);
+    /* va_list variant of the control function */
+    int (*vctrl)(struct ossl_event_queue_st *queue, uint32_t request,
+                 va_list args);
+
+/*
+ * Request the nummber of priorities for this method.
+ * Takes on extra argument:
+ * - a pointer to size_t, where the number of priorities will be stored
+ *
+ * This is a special request; it expects NULL for |queue|, and gives back a
+ * constant number.
+ */
+# define OSSL_EVENT_C_PRIORITIES    0x00000001
+};
+
+/*-
+ * Functions to create and populate events
+ * ---------------------------------------
+ */
+
+/* Allocate / deallocate an event structure dynamically and populate it */
+struct ossl_event_st *ossl_event_new(uint32_t type, void *ctx,
+                                     const void *identifiers,
+                                     void *payload, size_t payload_size,
+                                     ossl_event_destructor_fn *destructor);
+void ossl_event_free(struct ossl_event_st *);
+
+/*
+ * Populate an event structure.  Used for events that aren't allocated
+ * dynamically.
+ *
+ * Returns one of these values:
+ * 0    Not set
+ * 1    Set
+ */
+int ossl_event_set0(struct ossl_event_st *event,
+                    uint32_t type, void *ctx,
+                    const void *identifiers,
+                    void *payload, size_t payload_size,
+                    ossl_event_destructor_fn *destructor);
+
+/*
+ * Functions to create and populate event queues
+ * ---------------------------------------------
+ */
+
+/* Allocate / deallocate an event queue dynamically */
+struct ossl_event_queue_st *
+ossl_event_queue_new(const struct ossl_event_queue_method_st *method);
+void ossl_event_queue_free(struct ossl_event_queue_st *queue);
+
+/*
+ * Populate an event queue structure.  Used for event queues that aren't
+ * allocated dynamically.
+ *
+ * Returns one of these values:
+ * 0    Not set
+ * 1    Set
+ */
+int ossl_event_queue_set0(struct ossl_event_queue_st *queue,
+                          const struct ossl_event_queue_method_st *method);
+
+/*
+ * Subscribe / unsubscribe to events on an event queue
+ *
+ * Returns one of these values:
+ * -1   Error
+ *  0   Not found (unsubscibe only)
+ *  1   Subscribed / unsubscribed
+ */
+int ossl_event_queue_subscribe(struct ossl_event_queue_st *queue,
```

Comment:
Should a subscriber be able to specify what events they are interested in receiving?  Every subscriber getting every event seems like overkill.

I'm mostly thinking about how to streamline event dispatch to minimise the effort required.

Reply:
You mean the event type?  Yeah, I was thinking of that, but wonder if that would be much of a gain?  I imagined that they might want to filter on more than just that...  but sure, it wouldn't be hard to add a check of event type (and probably have a distinct event type that really means "any", for those that handle more than one event type)

Reply:
Minimising the number of events passed to call backs is *huge* benefit.

Reply:
Okie

Reply:
Ideally, a bit mask for the common classes of event (read, write, new connection, closing, alert/error, other) to make filtering very fast.